### ISSUE TITLE
unwind in ElasticSearch

### ISSUE THREAD

**Issue by gvijayakumar-16** [elastic/elasticsearch#28836]
Can we have an option in ElasticSearch, something similar to [unwind](https://docs.mongodb.com/manual/reference/operator/aggregation/unwind/) in MongoDB. 

A lot of relations exist between the data and querying becomes difficult.

**Comment by tlrx** [comment]
In which feature and how this `unwind` would be applicable?

**Comment by gvijayakumar-16** [comment]
Cases where I have a nested array of objects and I want the those objects to be queried, sorted and paginated.

**Comment by justinwyer** [comment]
Can you provide a more concrete example?

**Comment by gvijayakumar-16** [comment]
```
public class User
{
    public string Name { get; set; }
    //Properties
    List<Address> Addresses { get; set; }
}
```

Consider the above class which is the structure for ElasticSearch.

Now if I have a to display data based on the following scenarios:

By name:

 Name | Address
 -------|------------
 Adam|Chicago, Ohio

By Address

 Address| Name
 -------|------------
 Chicago|Adam
 Ohio|Adam

If there is a feature similar to unwind, getting the output for the second scenario would be much simpler.

Please note this is a simple example.

Even if I use reverse nested aggregation, pagination and sorting becomes a problem

**Comment by tlrx** [comment]
I understand the need for such a feature, but I think that this is something that must be done on the client side. Knowing how documents are streamed to build the response JSON I doubt that we could implement this feature easily without a lot of side effects. 

I'm going to mark this issue as "Discuss" so that it will be discussed in our weekly meetings.

**Comment by josefschiefer** [comment]
Unwind may help flatting the data for supporting simple query_string expressions with nested documents https://github.com/elastic/elasticsearch/issues/16551.

**Comment by jpountz** [comment]
We discussed it in Fixit Friday and agreed that we won't try to implement it due to the fact that we can't think of a way to support such operations efficiently.

The only ideas that we thought were reasonable boiled down to having another index that stores the same data but flattened. Depending on your use-case, you might be able to maintain those two views in parallel or would only maintain the one you have today, then materialize a flattened view of the data when you need it and throw it away after you are done querying. In both cases, this requires client-side logic.

**Comment by SecretPocketCat** [comment]
Are there any new features which would help me achieve uwnind followed by a sort as of 7.4 or am I better off duplicating the data across multiple indices?

E.g. There's a _Project_ index with the nested collections _Comments_ and _Makes_ and I'd like to filter the root _Project_ documents as well as the  _Comments_ and _Makes_, each with its own filter, unwind the hits & sort it by a date from each document.

**Comment by jpountz** [comment]
Maybe try denormalizing the data at index-time? Or maybe inner hits would help? https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-body.html#request-body-search-inner-hits

**Comment by SecretPocketCat** [comment]
> Maybe try denormalizing the data at index-time? 

As in creating different index/indices?

> Or maybe inner hits would help? https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-body.html#request-body-search-inner-hits

I need a 'global sort' based on each hit, which includes inner hits. I can't really sort by each inner hit, can I? That's where the unwind would come in.  

What I meant was whether there is some new feature (e.g. some new type of aggregation) that would cover hit/innerhit unwind and was added since the issue was closed. I skimmed the docs and have not encountered anything of interest, but I I'm not sure that I didn't miss something, hence the necro of this issue.
